buildSrc/src/main/java/org/elasticsearch/gradle/internal/InternalBwcGitPlugin.java

```
@@ -0,0 +1,202 @@
+     * after the fact e.x. a merge done today can add commits dated with yesterday so the result will no longer be
+     * deterministic.
+     * <p>
+     * We don't use merge commits, but for additional safety we check that no such commits exist in the time period
+     * we are interested in.
+     * <p>
+     * Timestamps are at seconds resolution. rev-parse --before and --after are inclusive w.r.t the second
+     * passed as input. This means the results might not be deterministic in the current second, but this
+     * should not matter in practice.
+     */
+    private String maybeAlignedRefSpec(Project project, Logger logger, String defaultRefSpec) {
+        if (project.getProviders().systemProperty("bwc.checkout.align").isPresent() == false) {
+            return defaultRefSpec;
+        }
+
+        String timeOfCurrent = execGit(execSpec -> {
+            execSpec.commandLine(asList("git", "show", "--no-patch", "--no-notes", "--pretty='%cD'"));
+            execSpec.workingDir(project.getRootDir());
+        });
+
+        logger.lifecycle("Commit date of current: {}", timeOfCurrent);
+
+        String mergeCommits = execGit(
+            spec -> spec.commandLine(asList("git", "rev-list", defaultRefSpec, "--after", timeOfCurrent, "--merges"))
+        );
+        if (mergeCommits.isEmpty() == false) {
+            throw new IllegalStateException("Found the following merge commits which prevent determining bwc commits: " + mergeCommits);
+        }
+        return execGit(
+            spec -> spec.commandLine(asList("git", "rev-list", defaultRefSpec, "-n", "1", "--before", timeOfCurrent, "--date-order"))
+        );
+    }
+
+    private void writeFile(File file, String content) {
+        try {
+            FileWriter myWriter = new FileWriter(file, false);
+            myWriter.write(content);
+            myWriter.close();
+        } catch (IOException e) {
+            e.printStackTrace();
```

Let's not swallow any exceptions. This should cause the build to fail. If we can't let it propagate as is, then wrap it with `UncheckedIOException`?

---

fixed